receiver/prometheusreceiver/internal/metrics_adjuster.go

```diff
@@ -42,55 +102,102 @@ func getSignature(name string, values []*metricspb.LabelValue) string {
 	return fmt.Sprintf("%s,%s", name, strings.Join(labelValues, ","))
 }
 
-// JobsMap maps from a job instance to a map of metric instances for the job.
-type JobsMap map[string]*metricsInstanceMap
+// JobsMap maps from a job instance to a map of timeseries instances for the job.
+type JobsMap struct {
+	sync.RWMutex
+	gcInterval time.Duration
+	lastGC     time.Time
+	jobsMap    map[string]*timeseriesMap
+}
 
 // NewJobsMap creates a new (empty) JobsMap.
-func NewJobsMap() *JobsMap {
-	jm := JobsMap(make(map[string]*metricsInstanceMap))
-	return &jm
+func NewJobsMap(gcInterval time.Duration) *JobsMap {
+	return &JobsMap{gcInterval: gcInterval, lastGC: time.Now(), jobsMap: make(map[string]*timeseriesMap)}
 }
 
-func (jm *JobsMap) get(job, instance string) *metricsInstanceMap {
+// remove
+func (jm *JobsMap) gc() {
+	current := time.Now()
+	jm.Lock()
+	// recheck lastGC to make sure gc is necessary
+	if current.Sub(jm.lastGC) > jm.gcInterval {
+		for sig, tsm := range jm.jobsMap {
+			tsm.Lock()
+			if !tsm.mark {
+				delete(jm.jobsMap, sig)
+			} else {
+				tsm.gc()
+			}
+			tsm.Unlock()
+		}
+		jm.lastGC = time.Now()
+	}
+	jm.Unlock()
```

Comment:
Yes, I can move the tsm.lock() to tsm.gc().

It is is possible that the metricsadjuster could mark the tsm and it will be removed from the JobsMap when it didn't need to be but that would happen even with locking (in a slightly different order).